File: validParentheses/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        m = {
            "(": ")",
            "[": "]",
            "{": "}",
        }
        for ss in s:
            if ss in {"(", "{", "["}:
                stack.append(ss)
            else:
                try:
                    last = stack.pop()
                    if m[last] != ss:
                        return False
                except IndexError:
                    return False
        return True if not stack else False

File: validParentheses/valid_parentheses.go
func isValid(s string) bool {
	stack := []rune{}
	m := map[rune]rune{
		'(': ')',
		'[': ']',
		'{': '}',
	}
	for _, char := range s {
		_, ok := m[char]
		if ok {
			stack = append(stack, char)
		} else {
			if len(stack) == 0 {
				return false
			}

			last := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			val, _ := m[last]
			if val != char {
				return false
			}
		}
	}

	if len(stack) > 0 {
		return false
	}
	return true
}
